File: ios/Classes/Channel/Methods/WalletStripeInitHandler.swift
//
//  WalletStripeInitHandler.swift
//
//
//  Created by Miguel Ángel Soto González on 20/10/25.
//

import Flutter

class WalletStripeInitHandler {
    let stripe: StripeManagerProtocol

    init(stripe: StripeManagerProtocol = StripeManager.shared) {
        self.stripe = stripe
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let options = arguments["options"] as? [String: Any],
              let publishableKey = options["publishableKey"] as? String,
              let merchantName = options["merchantName"] as? String
        else {
            let error = WalletStripeError.missingArgument(method: WalletStripeMethodType.initialize.rawValue)
            result(error.rawValue)
            return
        }

        let allowsDelayedPaymentMethods = options["allowsDelayedPaymentMethods"] as? Bool ?? false
        let primaryButtonColorHex = options["primaryButtonColor"] as? Int

        stripe.initialize(publishableKey: publishableKey,
                          merchantName: merchantName,
                          allowsDelayedPaymentMethods: allowsDelayedPaymentMethods,
                          primaryButtonColorHex: primaryButtonColorHex)
        result(nil)
    }
}

File: android/src/main/kotlin/com/rudo/hybrid_stripe_payments/channel/methods/WalletStripeInitHandler.kt
package com.rudo.hybrid_stripe_payments.channel.methods

import com.rudo.hybrid_stripe_payments.channel.error.WalletStripeError
import com.rudo.hybrid_stripe_payments.core.stripe.StripeManager
import com.rudo.hybrid_stripe_payments.core.stripe.StripeManagerInterface
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import kotlin.collections.get

class WalletStripeInitHandler(
    private val stripeManager: StripeManagerInterface = StripeManager
)  {

    private companion object {
        const val OPTIONS = "options"
        const val PUBLISHABLE_KEY = "publishableKey"
        const val MERCHANT_NAME = "merchantName"
        const val ALLOWS_DELAYED_PAYMENT_METHODS = "allowsDelayedPaymentMethods"
        const val PRIMARY_BUTTON_COLOR = "primaryButtonColor"

    }

    fun handle(call: MethodCall, result: Result) {
        val arguments = call.arguments as? Map<*, *>
        val options = arguments?.get(OPTIONS) as? Map<*, *>

        val publishableKey = options?.get(PUBLISHABLE_KEY) as? String
        val merchantName = options?.get(MERCHANT_NAME) as? String

        if (publishableKey == null || merchantName == null) {
            val walletError = WalletStripeError.MissingArgument(WalletStripeMethodType.Initialize.rawValue)
            result.error(walletError.code(), walletError.message(), walletError.details())
            return
        }

        val allowsDelayedPaymentMethods = options[ALLOWS_DELAYED_PAYMENT_METHODS] as? Boolean ?: false
        val primaryButtonColor = (options[PRIMARY_BUTTON_COLOR] as? Number)?.toInt()

        stripeManager.initialize(
            publishableKey = publishableKey,
            merchantName = merchantName,
            allowsDelayedPaymentMethods = allowsDelayedPaymentMethods,
            primaryButtonColor = primaryButtonColor
        )

        result.success(null)
    }
}
